modules/gapi/test/common/gapi_video_tests_inl.hpp

```diff
@@ -131,6 +131,234 @@ TEST_P(BackgroundSubtractorTest, AccuracyTest)
+
+        if (haveMeasure)
+            cv::randu(measure_vec, Scalar::all(-1), Scalar::all(1));
+
+        // G-API
+        comp.apply(cv::gin(measure_vec, haveMeasure), cv::gout(gapiKState));
+
+        // OpenCV
+        ocvKState = ocvKalman.predict();
+        if (haveMeasure)
+            ocvKState = ocvKalman.correct(measure_vec);
+    }
+
+    // Comparison //////////////////////////////////////////////////////////////
+    {
+        double diff = 0;
+        vector<int> idx;
+        EXPECT_TRUE(cmpEps(gapiKState, ocvKState, &diff, 1.0, &idx, false) >= 0);
+    }
+}
+
+TEST_P(KalmanFilterCircleSampleTest, AccuracyTest)
+{
+    // auxiliary variables
+    cv::Mat processNoise(2, 1, type);
+
+    // Input mesurement
+    cv::Mat measurement = Mat::zeros(1, 1, type);
+    // Angle and it's delta(phi, delta_phi)
+    cv::Mat state(2, 1, type);
+
+    // G-API graph initialization
+    cv::gapi::video::KalmanParams kp;
+
+    kp.state = Mat::zeros(2, 1, type);
+    randn(kp.state, Scalar::all(0), Scalar::all(0.1));
+
+    kp.errorCov = Mat::eye(2, 2, type);
+
+    if (type == CV_32F)
+        kp.transitionMatrix = (Mat_<float>(2, 2) << 1, 1, 0, 1);
+    else
+        kp.transitionMatrix = (Mat_<double>(2, 2) << 1, 1, 0, 1);
+
+    kp.processNoiseCov = Mat::eye(2, 2, type) * (1e-5);
+    kp.measurementMatrix = Mat::eye(1, 2, type);
+    kp.measurementNoiseCov = Mat::eye(1, 1, type) * (1e-1);
+
+    cv::GMat m;
+    cv::GOpaque<bool> have_mesure;
+    cv::GMat out = cv::gapi::KalmanFilter(m, have_mesure, kp);
+    cv::GComputation comp(cv::GIn(m, have_mesure), cv::GOut(out));
+
+    // OCV Kalman initialization
+    KalmanFilter KF(2, 1, 0);
+    initKalmanFilter(kp, KF, false);
+
+    randn(state, Scalar::all(0), Scalar::all(0.1));
+
+    // GAPI Corrected state
+    cv::Mat gapiCorrState(2, 1, type);
+    // OCV Corrected state
+    cv::Mat ocvCorrState(2, 1, type);
+    // GAPI Predicted state
+    cv::Mat gapiPreState(2, 1, type);
+    // OCV Predicted state
+    cv::Mat ocvPreState(2, 1, type);
+
+    bool haveMeasure;
+
+    for (int i = 0; i < numIter; ++i)
+    {
+        // Get GAPI Prediction
+        haveMeasure = false;
+        comp.apply(cv::gin(measurement, haveMeasure), cv::gout(gapiPreState));
+
+        // Get OCV Prediction
+        ocvPreState = KF.predict();
+
+        GAPI_DbgAssert(cv::norm(kp.measurementNoiseCov, KF.measurementNoiseCov, cv::NORM_INF) == 0);
+        // generation measurement
+        randn(measurement, Scalar::all(0), Scalar::all((type == CV_32F) ?
+              kp.measurementNoiseCov.at<float>(0) : kp.measurementNoiseCov.at<double>(0)));
+
+        GAPI_DbgAssert(cv::norm(kp.measurementMatrix, KF.measurementMatrix, cv::NORM_INF) == 0);
+        measurement += kp.measurementMatrix*state;
+
+        if (theRNG().uniform(0, 4) != 0)
+        {
+            haveMeasure = true;
+            KF.correct(measurement);
+            comp.apply(cv::gin(measurement, haveMeasure), cv::gout(gapiCorrState));
```

Comment:
Are you sure G-API code works as OCV?
You are calling `apply` above with `haveMeasure = false` which will cause the call of `predict` in our kernel. And here you're calling `apply` one more time so `predict` will be called one more time, whether `haveMeasure = true` or not.
But OCV KalmanFilter always calls `predict` only once.

Reply:
Ok. Reworked.